server/src/main/java/org/apache/druid/segment/realtime/FireHydrant.java

```diff
@@ -34,17 +34,53 @@
 
 import javax.annotation.Nullable;
 import java.io.Closeable;
+import java.io.File;
 import java.util.Optional;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Function;
 
 /**
+ *
  */
 public class FireHydrant
 {
   private final int count;
   private final AtomicReference<ReferenceCountingSegment> adapter;
   private volatile IncrementalIndex index;
+  private File persistedFile;
+  private SegmentId persistedSegmentId;
+
+  /**
+   * @return The persisted file path. This is needed to recreate mapped files before merging.
+   * it will be null for real time hydrants.
+   */
+  public @Nullable File getPersistedFile()
+  {
+    return persistedFile;
+  }
+
+  /**
+   * @return The persisted segment id. This is needed to recreate mapped files before merging.
+   * It will be null for real time hydrants
+   */
+  public @Nullable SegmentId getPersistedSegmentId()
```

Comment:
can you also add a note about why `persistedSegmentId` is required in addition to `getSegmentId()`? I understand that the latter will throw an NPE?

Reply:
Having `persistedSegmentId` in addition to `segmentId` is kind of ugly. But I was trying to avoid touching the existing code as much as possible. In order to use only segment id then I would have to introduce a "segmentId" private member that sometimes it is set by the adapter and others it is set by the "persisted segment id". I am on the fence on whether doing this would even be more confusing still. In any case, our discussion was for the next step to do is to split Appendarator into two implementations: One batch and one for real time. Then this code is temporary and we can have a cleaner implementation in the BatchAppenderatorImpl.

Reply:
I don't understand what `persistedSegmentId` is. Is it ever different from `segmentId`?

Reply:
The `persistedSegmentId` is the same as `segmentId`. However, `segmentId` is being accessed indirectly through the adapter. Closing the `QueryableIndex` nullifies the reference to it inside the adapter thus after this action `segmentId` is no longer accesible. I decided to add a new data member and a new method to store the `segmentId` after the queryable index is closed. The `segmentId` is required to re-open the `QueryableIndex` in the merge phase.

Reply:
In order to make the code less confusing I can still keep the `persistedSegmentId` reference but modify the `getSegmentId` method as follows:
`  public SegmentId getSegmentId()
  {
    if (adapter.get() != null) {
      return adapter.get().getId();
    } else {
      return persistedSegmentId;
    }
  }`

Reply:
Moved persisted metadata to `AppenderatorImpl` as a map. Added comments to the Map to why it is necessary. I think this way is much cleaner. In next push.